### ISSUE TITLE
Optimize Pod informer memory efficiency used in admission plugins

### ISSUE THREAD

**Issue by linxiulei** [kubernetes/kubernetes#125469]
### What would you like to be added?

Memory optimizations on Pod informer used by admission plugins to reduce memory overhead of apiserver.

Currently, three admission plugins are using Pod informer:

1. ServiceAccount
2. NodeRestriction
3. PodSecurity

Optimization strategies:

For ServiceAccount & NodeRestriction can be straight-forward as they only need Metadata fields and `pod.Spec.NodeName`, so once trim off other unused fields, the memory overhead can be minimized.

For PodSecurity, it's a bit more complicated as most fields of Pod object are used for evaluation. However, PodSecurity is not enabled or enabled with default mode (i.e. privileged), which doesn't require caching Pod objects. So the idea is to only enable/start informer when a namespace is configured with non-privileged mode. This might need further thinking with negative impact such as the latency of starting an informer etc.

### Why is this needed?

As a parallel effort to https://github.com/kubernetes/kubernetes/pull/121979, this work would reduce memory overhead of apiserver since the Pod objects are commonly the largest memory overhead of informer object cache.

**Comment by liggitt** [comment]
I don't see a path to not watching full pods.

Node authorization uses a graph populated by extracting all objects referenced by pods, which requires visibility to the full pod including all pod fields.

**Comment by linxiulei** [comment]
@liggitt thank you for pointing out node authorization. It's certainly trickier but my understanding is Node Authorization only needs the full Pod object when there are changes and once the graph is built the Pod objects can be discarded (i.e. local Pod object cache is never directly accessed but via EventHandler). So if we have a way to only watch the Pod resource without caching them locally or only caching the index key, we can still make it happen.

**Comment by puneetdixit200** [comment]
Hi maintainers, I’d like to work on this. I can start by mapping the current admission plugin informer usage, measure memory impact, and submit a targeted optimization plan/PR.

**Comment by k8s-1** [comment]
I'll work on a PR.

**Comment by k8s-1** [comment]
I've been looking into this deeper.

### Findings

NodeRestriction and ResourceQuota admission are both on by default, and both already read pods off the same shared informer. So that cache runs full-time on any normal cluster, regardless of what PodSecurity does.

ResourceQuota (`pkg/quota/v1/evaluator/core/registry.go`) reads pods too, which hasn't been mentioned yet.

Given both of those, I don't think the PodSecurity-only version would move the needle much on its own.

### Node authorizer

@liggitt raised that node authorization needs full pod visibility. I think that's true for the node authorizer's event stream, but not for a cached store.

It only reads pods via `AddEventHandler`, copies what it needs into its own graph, and drops the pod. It never calls `.Lister()`. So it could get its own dedicated watch instead of sharing the admission plugins' informer, without losing visibility.

Let me know if I'm misreading how that's used elsewhere.

### PodSecurity

PodSecurity is also a blocker for trimming the shared cache, separately from the authorizer. Its policy evaluation touches most of the pod spec.

It only reads pods in one place though (`EvaluatePodsInNamespace`, which fires when a namespace's enforce level changes), so it doesn't strictly need a cache at all. There's an unused live-client `PodLister` (`PodListerFromClient`) sitting right next to the informer-backed one that's actually wired up.

### Proposed direction

Three individually-safe PRs:

1. Give the node authorizer its own dedicated watch: no behavior change, just decouples it from the shared cache (see #141381 ).
2. Switch PodSecurity to the existing `PodListerFromClient`: no behavior change, removes it from the shared cache too.
3. Once both land, trim what the shared informer caches down to what NodeRestriction and ResourceQuota actually use.

> [!NOTE]
> Opened a draft PR for step 1: #141381. Wanted the approach in front of people before going further. It's not memory-positive by itself (details in the PR description), just removes one of the two things blocking step 3.

Does this direction seem reasonable, or is there something about the node authorizer / PodSecurity usage I'm not accounting for?

**Comment by liggitt** [comment]
> It only reads pods via AddEventHandler, copies what it needs into its own graph, and drops the pod. It never calls .Lister()

Doing that processing in the Add event handler is actually slower than it should be in high volume clusters. We're actively working on switching that to a normal queue / worker, which *would* use the pods from the lister in the worker (xref #140803)

**Comment by k8s-1** [comment]
> > It only reads pods via AddEventHandler, copies what it needs into its own graph, and drops the pod. It never calls .Lister()
> 
> Doing that processing in the Add event handler is actually slower than it should be in high volume clusters. We're actively working on switching that to a normal queue / worker, which _would_ use the pods from the lister in the worker (xref [#140803](https://github.com/kubernetes/kubernetes/pull/140803))

Good catch, missed #140803. I was wrong that giving the authorizer its own informer would still save memory. The queue in #140803 needs .Get(ns, name), which means the authorizer's own informer needs a real indexed store, not just an event stream. So instead of the authorizer costing nothing, it'd keep a full-pod copy of its own on top of the shared trimmed one.

So PR1 as I'd pitched it doesn't hold up on its own. I flagged the conflict on #140803.

**Comment by michaelasp** [comment]
> Good catch, missed https://github.com/kubernetes/kubernetes/pull/140803. I was wrong that giving the authorizer its own informer would still save memory. The queue in https://github.com/kubernetes/kubernetes/pull/140803 needs .Get(ns, name), which means the authorizer's own informer needs a real indexed store, not just an event stream. So instead of the authorizer costing nothing, it'd keep a full-pod copy of its own on top of the shared trimmed one.

Yeah and we do need certain fields like `pod.Spec.ServiceAccountName` in that. Especially for cases like node authorizer where we do expensive operations, it's not great to block at the event stream since that holds up other listeners. I'd look at seeing if we can trim down fields, such as with a transforming store.

**Comment by michaelasp** [comment]
Also looking at the proposal

> It only reads pods in one place though (EvaluatePodsInNamespace, which fires when a namespace's enforce level changes), so it doesn't strictly need a cache at all. There's an unused live-client PodLister (PodListerFromClient) sitting right next to the informer-backed one that's actually wired up.

I'd be cautious about live listers. That adds load to the apiserver, we should see how much that takes in certain envs.

**Comment by k8s-1** [comment]
@michaelasp 

**On the transforming-store idea**
Half of it's already done. ManagedFields has been stripped off this exact shared informer since 2024 (pkg/controlplane/apiserver/config.go), before this issue opened.

I went through the rest of Status field by field against the four things reading off this cache. Most of it's needed by someone (ResourceQuota's resize-aware accounting, the node authorizer's DRA claim edges). What's left unused is a small batch. Measured on kind, it's ~200 bytes/pod normally, ~800 only during an active debug session. A tenth of what ManagedFields was already worth, and there's no compiler check tying the trim to what these four actually read. If one of them starts reading a dropped field later, nothing flags it, it just quietly breaks.

**On the live-lister concern**
I don't have load data for that. Turned out not to matter much for the bigger picture though: even with PodSecurity fully off this cache, the authorizer's graph.AddPod already forces most of Spec to stay cached, so that move alone wouldn't have opened up much more room anyway.

**Comment by k8s-1** [comment]
@michaelasp @liggitt I've opened a draft for the field-trim idea: #141477.

This extends the existing ManagedFields trim to also clear a batch of Spec fields none of the four consumers here read (full list in the PR). Measured ~9-15% smaller cached Pod objects on kind, depending on which scheduling fields a pod uses. That's per-object in this specific cache, not a claim about total apiserver memory, I haven't measured that.

There's a tradeoff: no compiler check ties the field list to what the four consumers actually read.

**Comment by michaelasp** [comment]
I'm generally not a huge fan of transforms wiping fields in spec/status since if they're ever used in the future then we open ourselves up to bugs if a future author is unaware of the transform and assumes the field is populated. The cilium project does this by adding a slim_v1 wrapper on top of corev1 https://github.com/cilium/cilium/blob/main/pkg/k8s/slim/README.md so we never can access wiped fields but not sure if we'd want to do something of the kind in k/k like that.

**Comment by Arman16-1998** [comment]
Hi! I’m interested in working on this issue. I’d like to investigate the Pod informer memory usage in the admission plugins and understand where the current duplication/overhead comes from. Since this issue is unassigned and triage/accepted, is it okay if I take it?
